src/parsec.rs

```
@@ -658,26 +659,19 @@ impl<T: NetworkEvent, S: SecretId> Parsec<T, S> {
             return Ok(PostProcessAction::Continue);
         }
 
-        let mut peer_list_changes = Vec::with_capacity(payload_keys.len());
-        let mut blocks = Vec::with_capacity(payload_keys.len());
-
-        for payload_key in &payload_keys {
-            self.output_consensus_info(payload_key);
-
-            match self.create_block(&payload_key) {
-                Ok(block) => blocks.push(block),
-                Err(Error::MissingVotes) => (),
-                Err(error) => return Err(error),
-            }
-
-            self.mark_observation_as_consensused(&payload_key);
-            peer_list_changes.extend(self.handle_consensus(payload_key));
-        }
+        self.output_consensus_info(&payload_keys);
+        self.mark_observations_as_consensused(&payload_keys);
```

Why do we mark it before when we use to mark it after?

---

The order doesn't matter.

---

Should this be handled as:
```
let peer_list_changes = payload_keys
            .iter()
            .filter_map(|payload_key| {
                 self.mark_observation_as_consensused(payload_key);
                 self.handle_consensus(payload_key)
             })
```

---

We can do that, but there is no functional difference.

---

The order matter in case of error in create_block.
This could well be that we want to mark observation as consensused even if create_blocks error or filter them out, but it may warrant a comment.

---

I see what you mean now. I guess I can move `mark_observations_as_consensused` after the blocks creation. That should be cleaner, no? I'll also adress @Fraser999 's nit while I'm at it.

---

The remaining question, is what happen if some were filtered?
Before they would not be marked as consensused or call to 
handle_consensus.

I guess it may not happen since missing vote is not possible?
If that is the case, then should we not let the error up the stack instead of swallowing it, and no consensusing anything of these blocks?

---

If I remember correctly, the situation can happen when we are in `ConsensusMode::Single` and there is a opaque vote by someone who has subsequently been removed. So we would reach consensus on that vote, but there is no one to sign it anymore. We don't treat this situation as error, because it can legally happen. We still want to mark the vote as consensued, so the algorithm is not slowed down by having to constantly reprocess it.

---

Ok that make sense to me. (maybe a 1/2 line comment in the filter part to say that would make things clearer).

I guess this comes from:
`.filter(|event| voters.contains(event.creator()))`

(Not super clear why we should filter these out though, were still valid voters at the time they voted for the event (maybe) So we may drop valid event on the floor.)

---

I guess because we currently can't tell whether someone is not a voter because they haven't been promoted to one yet or because they have already been removed. Something we might want to change at some point, but I'd say it has low priority. Losing votes of people who are already gone is something we can live with for now.